flang/lib/Optimizer/Transforms/RewriteLoop.cpp

```diff
@@ -258,10 +258,23 @@ class CfgIterWhileConv : public mlir::OpRewritePattern<fir::IterWhileOp> {
 
     // With the body block done, we can fill in the condition block.
     rewriter.setInsertionPointToEnd(conditionBlock);
-    auto comp1 =
+    // The comparison depends on the sign of the step value. We fully expect
+    // this expression to be folded by the optimizer or LLVM. This expression
+    // is written this way so that `step == 0` always returns `false`.
+    auto zero = rewriter.create<mlir::ConstantIndexOp>(loc, 0);
+    auto compl0 =
+        rewriter.create<mlir::CmpIOp>(loc, CmpIPredicate::slt, zero, step);
+    auto compl1 =
         rewriter.create<mlir::CmpIOp>(loc, CmpIPredicate::slt, iv, upperBound);
```

Comment:
I believe this (unchanged) line was, and remains off by one for unit step cases and a subset of other cases.  The `slt` comparison should be `sle`.  And similarly on line 272.

Why is this different than the `DoLoopOp` case?  If this is as good or better than trip count control (doubtful), shouldn't the same be done there?

Reply:
Yes, that should be inclusive. Thanks.